clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -322,23 +333,26 @@ private boolean maybeSendTransactionalRequest(long now) {
                 }
                 if (targetNode != null) {
                     if (nextRequestHandler.isRetry()) {
+                        log.trace("waiting " + retryBackoffMs + "ms before resending a transactional request: " + nextRequestHandler.requestBuilder());
                         time.sleep(retryBackoffMs);
                     }
                     ClientRequest clientRequest = client.newClientRequest(targetNode.idString(), nextRequestHandler.requestBuilder(),
                             now, true, nextRequestHandler);
                     transactionManager.setInFlightRequestCorrelationId(clientRequest.correlationId());
+                    log.trace("sending transactional request {} to node {}", nextRequestHandler.requestBuilder(), clientRequest.destination());
                     client.send(clientRequest, now);
                     return true;
                 }
             } catch (IOException e) {
-                log.warn("Got an exception when trying to find a node to send a transactional request to. Going to back off and retry", e);
+                targetNode = null;
+                log.warn("Got an exception when trying to find a node to send transactional request " + nextRequestHandler.requestBuilder() + ". Going to back off and retry", e);
             }
+            log.trace("About to wait for " + retryBackoffMs + "ms before trying to send another transational request.");
             time.sleep(retryBackoffMs);
             metadata.requestUpdate();
         }
 
-        if (targetNode == null)
-            transactionManager.retry(nextRequestHandler);
+        transactionManager.retry(nextRequestHandler);
```

Comment:
Yes. The `networkClientUtils.awaitReady` and `awaitLeastLoadedNodeReady` throws an IOException if the connection to the node fails. So we would have a `targetNode` which is the coordinator, and then when we tried to connect to it, it would through an IOException. 

So `targetNode` would not be null, the transactional request would not be sent, and it wouldn't be reenqueued either.. so the top level methods like `commitTransaction`, `sendOffsets` etc., would wait forever.